### DISCUSSION TITLE
python的物件導向class學習心得

### DISCUSSION THREAD

**Discussion by Aiame** [1101-GS4719-C/Python-Community-GS4719-C#1458] [General]
Hi   我是個程式的新手，第一次發文還請見諒。想和大家分享class的內容，如果有錯誤還麻煩大家指教。
p.s. 大佬可以看看就好，和我一樣萌新也可以參考看看

為甚麼會想分享這個呢? 因為在我還沒那麼厲害的時候(幾乎0基礎)，系上的課讓我直接從c語言，跳到c++的資料結構，這真的是很新手不友善，阿老師上課又一直class來class去的，看得我眼花撩亂 QQ ，ok 正文開始。

物件(object)：物件是⼀種⾃訂的資料結構，裡⾯有資料(變數，稱為屬性)以及程式碼(函
式，稱為方法)。

類別(Class)：類別提供⼀種將資料與函數綁在⼀起的機制，將變數與函式封裝成為類別，封裝的變數與涵式通稱為類別的屬性
(from 老師的講義)

上面聽起來好饒口，那我們就用實例來介紹比較簡單。
假設今天我們要建立一個物件是手機好了，那我們的程式會長這樣

```  python
class phone:
    def __init__(self, os, color, waterproof):  #在我們建立一個物件時所需要給的初始值
        self.os = os        # 1
        self.color = color        #2   
        self.waterproof = waterproof        #3
#到這邊我們建立了一個class(類別)

iphone = phone("ios", silver, True)    #4
#直到這裡我們才建立了一個物件，叫做iphone
```

簡單來說，我們可以把class想像成一個設計藍圖，這個設計藍圖有以下這些資訊
``` python
#self. 的意思代表我的iphone要有的屬性，也就是我有個屬性叫 os ，他的內容是 ios
#有顏色是銀色，有防水(看#4)
self.os = os    
self.color = color
self.waterproof = waterproof
```

接著再到4.，這段的意思是我告訴工廠老闆說，我要做一台手機，這台手機的藍圖是這樣，其中有一些規格，到這裡就簡單介紹了class的作用，大家應該有一點懂了嗎，希望有...

然後可能有人會問，self 是個什麼東東，我明明規格有4個，阿輸入的怎麼只有三個(我也困惑一陣子)
這個 self. 的意思呢，是在告訴我說是哪一支手機有這個功能。
``` python
class phone:
    def __init__(self, os, color, waterproof):  #__init__是特別的語法，在我呼叫這個模板之後，就會自動生成裡面的東西
        self.os = os        
        self.color = color           
        self.waterproof = waterproof        

class phone2:
    def __init__(self, os, color):  
        self.os = os        
        self.color = color                 

iphone18 = phone()    #5

iphone19 = phone2()    #6
```
前面說到 class 像是設計圖，由上面的程式，我們就得到了兩支iphone
一隻18，一隻19，分別依照2種不同模板做出來
所以今天如果我想看 iphone18 的作業系統是什麼的時候，我就可以輸入
`print(iphone18.os)` 
這樣就能看到 iphone18 的作業系統哦( self = iphone18的意思，如果想換也可以，在#5把iphone18改掉就可以了
iphone19 不防水，是想趕快讓我換手機嗎?(怕.jpg

總結：
class的好處呢，我覺得在於它可以把程式寫得更有系統性，在class裡面還可以定義別的function(實體方法 instance methods)，那你可能會想說，我python可以用import的方式來寫啊，在另一個.py檔寫一堆 def ，再import進去不是一樣，但程式碼可能亂亂的，所以我們可以用class來把比較相似的程式碼封包在一起，讓程式寫的更好看。

接下來是上課老師出的一個小作業，我嘗試了一下，雖然還是有瑕疵，但這是我目前能做到最好的樣子
大家可以到我的repository看，之後有時間可能會再回來修改，程式裡本來有想用繼承的方法，但最後沒弄出來。

寫到這裡有點太長了，雖然我還有很多問題想問，就留到下次吧，之後會繼續做class或者ML的東西(ML好難 QQ)

reference：
1.https://www.youtube.com/@GrandmaCan
2.老師的講義
3.系上老師的講義
4.電神室友的建議

**Comment by alstinlin** [comment]
蠻認同的，因為對比較龐大的程式來說可讀性很重要，此外用class把類似功能的涵式分類也有助於後續的debug，總而言之，我也覺得class是一個很好用的工具~

**Comment by pudding0803** [comment]
自從對物件導向比較熟悉之後，我開始非常注重程式碼的寫法、專案整體的架構、以及檔案之間互相的關係。為了讓我能夠在未來較容易應對龐大的專案，我就算是小專案，也會盡可能讓可讀性越高越好。想當年一開始學程式，變數名稱都是 a、b、c、d 🥲

希望大家都可以多多嘗試並練習物件導向，一起讓未來的程式碼都變得更好看😇

**Comment by RogelioKG** [comment]
小弟是先入門 Python 的 OOP，接著再學 C / C++ 的 OOP。
如果有想要深入搞懂 OOP 的朋友，建議其實應該從 C++ 開始。
Python 有很多 OOP 概念都是沿用 C++ 的 (對 Java 不熟，有熟悉 Java 的朋友可補充🥹)。
例如 abc，一開始學真的搞不懂類別就類別為什麼要抽象，
後來才明白抽象類別是先形塑出一個介面 (interface)，等到後面再來覆寫 (override) 並實作 (implement)。
這是我在看完 C++ 後才體會到的理念。

**Reply by RogelioKG** [reply]
還有 @property、魔術方法、鴨子型別 (duck typing)、猴子補丁 (monkey patch) 這些蠻酷的 Python OOP 特色。
身邊蠻多朋友寫過程式，大多卻對物件導向沒有概念，真的有些可惜。
物件導向可以讓程式碼更有組織也更好維護，是一種值得擁有的設計思維。

**Comment by RogelioKG** [comment]
理解物件導向的概念後，也可以反過來幫助你在學習模組時，理解不同類別產生的物件與它們可使用的方法。
舉個栗子🌰，cv2 讀取一張圖片時物件的類別是 Mat，去翻 library 的 source code 就能發現 Mat 其實是型別別名 (type alias)
```py
Mat = np.ndarray[int, np.dtype[np.generic]]
```
因此你能明白 Mat 就是 numpy 模組中 ndarray 類別的特化型 (資料類別只能是 int)，
也能明白 ndarray 產生的物件所擁有的方法 (method) 能被 Mat 產生的物件所使用。

**Comment by lin157644** [comment]
雖然Python也允許procedure的方去去寫，但是它的物件導向設計真的相當完善，基本的繼承多型，也支援反射、多重繼承、抽象類別、Meta programming，整體設計邏輯也相當一致，若能妥善運用就可以寫出相當優雅的程式。
